### PR TITLE
Small compilation fixes for GCC 6.3

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Small compilation fixes with newer GCC on Ubuntu 17.04

### CODE REVIEW

**Comment by jai1** [comment]
@merlimat - Have you noticed an improvement in C++ client performance by moving to a higher gcc and boost version?

**Comment by merlimat** [comment]
@jai1 I haven't checked any performance, was just trying different compiling in newer Ubuntu

File: pulsar-client-cpp/lib/ObjectPool.h [hunk 1]
```diff
@@ -152,6 +152,7 @@ class Allocator {
     };
 
     static boost::thread_specific_ptr<Impl> implPtr_;
+    typedef T value_type;
```
Comment (jai1) [inline pulsar-client-cpp/lib/ObjectPool.h]: I don't understand the use of this change

Comment (merlimat) [inline pulsar-client-cpp/lib/ObjectPool.h]: Compiler was complaining that `value_type` was not defined

Comment (merlimat) [inline pulsar-client-cpp/lib/ObjectPool.h]: apparently that's a required trait for an allocator

**Pull request merged**